### DISCUSSION TITLE
Deseralizing a custom Vec class

### DISCUSSION THREAD

**Discussion by fluxehub** [stephenberry/glaze#501] [Q&A] [answered]
As part of a project, I've had to write my own basic vector math class, and I want to find some way to deseralize JSON directly to it. Reading the documentation, I see:

> Array types logically convert to JSON array values. Concepts are used to allow various containers and even user containers if they match standard library interfaces.

But I'm not sure what this looks like in practice (or what I would need to implement to support this). My Vec class at the moment looks like this:

```cpp
template<size_t N>
class Vec {
public:
    Vec() = default;

    explicit Vec(float v) {
        _values.fill(v);
    }

    Vec(const std::array<float, N> &values) : _values{values} {}

    template<typename... Args>
    explicit Vec(Args... args) : _values{args...} {}

    // ...
protected:
    std::array<float, N> _values;
};
```

and I'm trying to deseralize JSON like this (note the mix of int and float literals)
```
{
  "foo": [0, 1.0, 1]
  "bar": [0.3, -0.2, 0]
}
```

**Comment by stephenberry** [comment, marked as answer]
For a static vector like this, I think you just need to satisfy this concept:

```c++
template <class T>
   concept range = requires(T& t) {
                      requires !std::same_as<void, decltype(t.begin())>;
                      requires !std::same_as<void, decltype(t.end())>;
                      requires std::input_iterator<decltype(t.begin())>;
                   };

```

So, you need `begin` and `end` iterators to your vector.

You could just return the iterators from the std::array you are using.

**Reply by fluxehub** [reply]
That worked perfectly! Thanks for your help.
